sql/src/main/java/io/crate/protocols/postgres/types/TimestampType.java

```diff
@@ -87,13 +120,37 @@ public Object readBinaryValue(ByteBuf buffer, int valueLength) {
 
     @Override
     byte[] encodeAsUTF8Text(@Nonnull Object value) {
-        return ISO_FORMATTER.print((long) value).getBytes(StandardCharsets.UTF_8);
+        long msecs = (long) value;
+        if (msecs >= FIRST_MSEC_AFTER_CHRIST) {
+            return ISO_FORMATTER.print((long) value).getBytes(StandardCharsets.UTF_8);
+        } else {
+            return ISO_FORMATTER_WITH_ERA.print((long) value).getBytes(StandardCharsets.UTF_8);
+        }
     }
 
 
     @Override
     Object decodeUTF8Text(byte[] bytes) {
+        // Currently seems that only GoLang prepared statements are sent as TimestampType
+        // Other PostgreSQL clients send the parameter as Bigint or Varchar
         String s = new String(bytes, StandardCharsets.UTF_8);
-        return ISO_FORMATTER.parseMillis(s);
+
+        int endOfSeconds  = s.indexOf(".");
+        int idx = endOfSeconds;
+        if (endOfSeconds > 0) {
+            idx++;
+            while (s.charAt(idx) != '+' && s.charAt(idx) != '-') { // start of timezone
+                idx++;
+            }
+        }
+
+        int fractionDigits = idx - endOfSeconds- 1;
+        fractionDigits = fractionDigits < 0 ? 0 : fractionDigits;
+
+        boolean withEra = s.endsWith("BC") || s.endsWith("AD");
+        if (withEra) {
+            return PARSERS_WITH_ERA[fractionDigits].parseMillis(s);
+        }
+        return PARSERS_WITHOUT_ERA[fractionDigits].parseMillis(s);
```

Comment:
I think we should catch if there is no parser for fractionDigits  > 10 or?

Reply:
I wouldn't add this specific condition here. This parsing is very specific to what currently Go driver sends and if something is wrong a Runtime exception would be thrown and properly sent to the user anyway.

Reply:
sure but the exception wouldn't be really meaningful then...